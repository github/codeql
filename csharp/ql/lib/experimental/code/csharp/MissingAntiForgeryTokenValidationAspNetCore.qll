/** Library to support cs/web/missing-token-validation-aspnetcore query */

private import csharp
private import semmle.code.csharp.frameworks.system.Web
private import semmle.code.csharp.frameworks.system.web.Helpers as Helpers
private import semmle.code.csharp.frameworks.system.web.Mvc as Mvc
private import semmle.code.csharp.frameworks.microsoft.AspNetCore as AspNetCore

/** Defines classes used for anti-forgery tokens. */
class AntiForgeryClass extends Class {
  AntiForgeryClass() {
    this instanceof Helpers::AntiForgeryClass or
    this instanceof AspNetCore::AntiForgeryClass
  }

  /** Gets the `Validate` method. */
  Method getValidateMethod() {
    result = this.(Helpers::AntiForgeryClass).getValidateMethod() or
    result = this.(AspNetCore::AntiForgeryClass).getValidateMethod()
  }
}

/** Defines the anti-forgery token attribute. */
class ValidateAntiForgeryTokenAttribute extends Attribute {
  ValidateAntiForgeryTokenAttribute() {
    this instanceof Mvc::ValidateAntiForgeryTokenAttribute or
    this instanceof AspNetCore::ValidateAntiForgeryAttribute
  }
}

/** Defines a generic controller including Mvc and AspNetCore */
class Controller extends Class {
  Controller() {
    this instanceof Mvc::Controller or
    this instanceof AspNetCore::MicrosoftAspNetCoreMvcController
  }

  /** Gets a Method with a POST action */
  Method getAPostActionMethod() {
    result = this.(Mvc::Controller).getAPostActionMethod()
    or
    exists(Method method |
      method = this.(AspNetCore::MicrosoftAspNetCoreMvcController).getAnActionMethod() and
      method.getAnAttribute() instanceof AspNetCore::MicrosoftAspNetCoreMvcHttpPostAttribute and
      result = method
    )
  }
}

/** An `AuthorizationFilter` that calls the `AntiForgery.Validate` method. */
class AntiForgeryAuthorizationFilter extends Mvc::AuthorizationFilter {
  AntiForgeryAuthorizationFilter() {
    this.getOnAuthorizationMethod().calls*(any(AntiForgeryClass a).getValidateMethod())
  }
}

/** Find a filter that auto-validates anti-forgery tokens. */
class AutoValidateAntiForgeryTokenFilter extends Expr {
  AutoValidateAntiForgeryTokenFilter() {
    exists(
      MethodCall addControllers, LambdaExpr lambda, ParameterAccess options, PropertyCall filters,
      MethodCall add
    |
      // "AddMvc", "AddControllersWithViews", "AddMvcCore", "AddControllers", "AddRazorPages", so generalised to "Add*" to future-proof
      addControllers.getTarget().getName().matches("Add%") and
      addControllers.getArgument(1) = lambda and
      lambda.getAParameter().getAnAccess() = options and
      filters.getQualifier() = options and
      filters.getTarget().getName() = "get_Filters" and
      add.getQualifier() = filters and
      add.getTarget().getUndecoratedName() = "Add" and
      this = add and
      (
        // new AutoValidateAntiforgeryTokenAttribute()
        exists(ObjectCreation obj |
          add.getArgument(0) = obj and
          obj.getType() instanceof AutoValidateAntiforgeryTokenAttributeType
        )
        or
        // typeof(AutoValidateAntiforgeryTokenAttribute)
        exists(TypeAccess access |
          add.getArgument(0).(TypeofExpr).getAChild() = access and
          access.getType() instanceof AutoValidateAntiforgeryTokenAttributeType
        )
        or
        // Add<AutoValidateAntiforgeryTokenAttribute>()
        add.getTarget().getName() = "Add<AutoValidateAntiforgeryTokenAttribute>"
      )
    )
  }
}

/** Accounts for custom auto-validation classes with a similar name */
class AutoValidateAntiforgeryTokenAttributeType extends Type {
  AutoValidateAntiforgeryTokenAttributeType() {
    this.getName().matches("%AutoValidateAntiforgeryTokenAttribute")
  }
}

/** The ignore anti-forgery token attribute. */
class IgnoreAntiforgeryTokenAttribute extends Attribute {
  IgnoreAntiforgeryTokenAttribute() { this.getType().getName() = "IgnoreAntiforgeryTokenAttribute" }
}

/** The auto-validate anti-forgery token attribute. */
class AutoValidateAntiforgeryTokenAttribute extends Attribute {
  AutoValidateAntiforgeryTokenAttribute() {
    this.getType() instanceof AutoValidateAntiforgeryTokenAttributeType
  }
}

/**
 * Holds if the project has a global anti forgery filter.
 */
predicate hasGlobalAntiForgeryFilter() {
  // A global filter added
  exists(MethodCall addGlobalFilter |
    // addGlobalFilter adds a filter to the global filter collection
    addGlobalFilter.getTarget() = any(Mvc::GlobalFilterCollection gfc).getAddMethod() and
    // The filter is an antiforgery filter
    addGlobalFilter.getArgumentForName("filter").getType() instanceof AntiForgeryAuthorizationFilter and
    // The filter is added by the Application_Start() method
    any(WebApplication wa)
        .getApplication_StartMethod()
        .calls*(addGlobalFilter.getEnclosingCallable())
  )
  or
  // for ASP.NET Core
  exists(AutoValidateAntiForgeryTokenFilter filter)
}

/** Holds if the Method has the name "Login" */
predicate isLoginAction(Method m) { m.getName() = "Login" }

/** Holds if a Method has a CSRF attribute. */
predicate methodHasCsrfAttribute(Method method) {
  exists(Attribute attribute |
    (
      attribute instanceof ValidateAntiForgeryTokenAttribute or
      attribute instanceof IgnoreAntiforgeryTokenAttribute
    ) and
    (
      method.getAnAttribute() = attribute or
      method.getAnUltimateImplementee().getAnAttribute() = attribute
    )
  )
}

/** Holds if a Controller has a CSRF attribute. */
predicate controllerHasCsrfAttribute(Controller c) {
  exists(Attribute attribute |
    (
      attribute instanceof ValidateAntiForgeryTokenAttribute or
      attribute instanceof IgnoreAntiforgeryTokenAttribute or
      attribute instanceof AutoValidateAntiforgeryTokenAttribute
    ) and
    c.getBaseClass*().getAnAttribute() = attribute
  )
}
