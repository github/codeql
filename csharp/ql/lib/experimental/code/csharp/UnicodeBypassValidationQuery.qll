/**
 * Provides a taint-tracking configuration for detecting "Unicode transformation mishandling" vulnerabilities.
 */

private import csharp
import semmle.code.csharp.dataflow.internal.tainttracking1.TaintTracking
import semmle.code.csharp.security.dataflow.flowsources.Remote
import UnicodeBypassValidationCustomizations::UnicodeBypassValidation

/** A state signifying that a logical validation has not been performed. */
class PreValidation extends DataFlow::FlowState {
  PreValidation() { this = "PreValidation" }
}

/** A state signifying that a logical validation has been performed. */
class PostValidation extends DataFlow::FlowState {
  PostValidation() { this = "PostValidation" }
}

/**
 * A taint-tracking configuration for detecting "Unicode transformation mishandling" vulnerabilities.
 *
 * This configuration uses two flow states, `PreValidation` and `PostValidation`,
 * to track the requirement that a logical validation has been performed before the Unicode Transformation.
 */
class Configuration extends TaintTracking::Configuration {
  Configuration() { this = "UnicodeBypassValidation" }

  override predicate isSource(DataFlow::Node source, DataFlow::FlowState state) {
    source instanceof RemoteFlowSource and state instanceof PreValidation
  }

  override predicate isAdditionalTaintStep(
    DataFlow::Node nodeFrom, DataFlow::FlowState stateFrom, DataFlow::Node nodeTo,
    DataFlow::FlowState stateTo
  ) {
    (
      exists(MethodCall mc |
        mc.getTarget().hasName("Escape") and
        mc.getTarget()
            .getDeclaringType()
            .getABaseType*()
            .hasQualifiedName("System.Security", "SecurityElement") and
        nodeFrom.asExpr() = mc.getArgument(0) and
        nodeTo.asExpr() = mc
      )
      or
      exists(MethodCall mc |
        mc.getTarget()
            .hasName([
                "Count", "Equals", "IsMatch", "Match", "UseOptionC", "UseOptionR",
                "ValidateMatchTimeout"
              ]) and
        mc.getTarget()
            .getDeclaringType()
            .getABaseType*()
            .hasQualifiedName("System.Text.RegularExpressions", "Regex") and
        nodeFrom.asExpr() = mc.getArgument(0) and
        nodeTo = nodeFrom
      )
      or
      exists(MethodCall mc |
        mc.getTarget()
            .hasName(["Matches", "Replace", "Split", "Unescape", "Escape", "EnumerateMatches",]) and
        mc.getTarget()
            .getDeclaringType()
            .getABaseType*()
            .hasQualifiedName("System.Text.RegularExpressions", "Regex") and
        nodeFrom.asExpr() = mc.getArgument(0) and
        nodeTo.asExpr() = mc
      )
      or
      exists(MethodCall mc |
        mc.getTarget()
            .hasName([
                "Clone", "Compare", "CompareOrdinal", "CompareTo", "Concat", "Contains", "Copy",
                "CopyTo", "Create", "EndsWith", "EnumerateRunes", "Equals", "Format",
                "GetEnumerator", "GetHashCode", "GetPinnableReference", "GetTypeCode", "IndexOf",
                "IndexOfAny", "Insert", "Intern", "IsInterned", "IsNormalized", "IsNullOrEmpty",
                "IsNullOrWhiteSpace", "Join", "LastIndexOf", "LastIndexOfAny", "Normalize",
                "PadLeft", "PadRight", "Remove", "Replace", "ReplaceLineEndings", "Split",
                "StartsWith", "Substring", "ToCharArray", "ToLower", "ToLowerInvariant", "ToString",
                "ToUpper", "ToUpperInvariant", "Trim", "TrimEnd", "TrimStart", "TryCopyTo"
              ]) and
        mc.getTarget().getDeclaringType().getABaseType*().hasQualifiedName("System", "String") and
        nodeFrom.asExpr() = mc.getQualifier() and
        nodeTo = nodeFrom
      )
    ) and
    stateFrom instanceof PreValidation and
    stateTo instanceof PostValidation
  }

  /* A Unicode Tranformation (Unicode tranformation) is considered a sink when the algorithm used is either NFC or NFKC.  */
  override predicate isSink(DataFlow::Node sink, DataFlow::FlowState state) {
    exists(MethodCall mc, MemberConstant tm, MemberConstantAccess mca |
      mca = mc.getArgument(0) and
      mc.getTarget().hasName("Normalize") and
      tm = mca.getTarget() and
      tm.toString() = ["FormKC", "FormC"] and
      tm.getType().toString() = "NormalizationForm" and
      sink.asExpr() = mc.getQualifier()
    ) and
    state instanceof PostValidation
  }
}
