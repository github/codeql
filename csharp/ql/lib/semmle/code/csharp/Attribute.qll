/**
 * Provides the `Attribute` class.
 */

import Type
private import semmle.code.csharp.ExprOrStmtParent
private import TypeRef

/**
 * An element that can have attributes. Either an assembly (`Assembly`), a field (`Field`),
 * a parameter (`Parameter`), an operator (`Operator`), a method (`Method`), a constructor (`Constructor`),
 * a destructor (`Destructor`), a callable accessor (`CallableAccessor`), a value or reference type
 * (`ValueOrRefType`), a declaration with accessors (`DeclarationWithAccessors`), a local function
 * (`LocalFunction`) or a lambda expression (`LambdaExp`).
 */
class Attributable extends @attributable {
  /** Gets an attribute attached to this element, if any. */
  final Attribute getAnAttribute() { result.getTarget() = this }

  /** Gets a textual representation of this element. */
  string toString() { none() }

  /**
   * Holds if this element is at the specified location.
   * The location spans column `startcolumn` of line `startline` to
   * column `endcolumn` of line `endline` in file `filepath`.
   * For more information, see
   * [Locations](https://codeql.github.com/docs/writing-codeql-queries/providing-locations-in-codeql-queries/).
   */
  predicate hasLocationInfo(
    string filepath, int startline, int startcolumn, int endline, int endcolumn
  ) {
    this.(Element)
        .getLocation()
        .hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
  }
}

private string getAttributeName(Attribute a) {
  exists(string type, string pattern |
    type = a.getType().getName() and pattern = "(.*)Attribute(<.*>)?"
  |
    type.regexpMatch(pattern) and
    result = concat(int i | i = [1, 2] | type.regexpCapture(pattern, i) order by i)
    or
    not type.regexpMatch(pattern) and
    result = type
  )
}

/**
 * An attribute, for example `[...]` on line 1 in
 *
 * ```csharp
 * [DllImport("kernel32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
 * public static extern int GetFinalPathNameByHandle(
 *   SafeHandle handle,
 *   [In, Out] StringBuilder path,
 *   int bufLen,
 *   int flags);
 * ```
 */
class Attribute extends TopLevelExprParent, @attribute {
  /** Gets the type of this attribute. */
  Class getType() { attributes(this, _, getTypeRef(result), _) }

  /** Gets the element that this attribute is attached to. */
  Attributable getTarget() { attributes(this, _, _, result) }

  /**
   * Gets the `i`th argument of this attribute. This includes both constructor
   * arguments and named arguments.
   */
  Expr getArgument(int i) { result = this.getChildExpr(i) }

  /**
   * Gets the `i`th constructor argument of this attribute. For example, only
   * `true` is a constructor argument in
   *
   * ```csharp
   * MyAttribute[true, Foo = 0]
   * ```
   */
  Expr getConstructorArgument(int i) {
    result = this.getArgument(i) and not exists(result.getExplicitArgumentName())
  }

  /**
   * Gets the named argument `name` of this attribute. For example, only
   * `0` is a named argument in
   *
   * ```csharp
   * MyAttribute[true, Foo = 0]
   * ```
   */
  Expr getNamedArgument(string name) {
    result = this.getArgument(_) and
    result.getExplicitArgumentName() = name
  }

  override Location getALocation() { attribute_location(this, result) }

  override string toString() { result = "[" + getAttributeName(this) + "(...)]" }

  override string getAPrimaryQlClass() { result = "Attribute" }
}

/**
 * A generic attribute, for example `[...]` on line 1 in
 *
 * ```csharp
 * [MyGenericAttribute<int>(0)]
 * public void SomeMethod(string s) { }
 * ```
 */
class GenericAttribute extends Attribute {
  private ConstructedClass type;

  GenericAttribute() { type = this.getType() }

  /** Gets the total number of type arguments. */
  int getNumberOfTypeArguments() { result = count(int i | type_arguments(_, i, type)) }

  /** Gets the `i`th type argument, if any. */
  Type getTypeArgument(int i) { result = type.getTypeArgument(i) }

  /** Get a type argument. */
  Type getATypeArgument() { result = this.getTypeArgument(_) }

  override string getAPrimaryQlClass() { result = "GenericAttribute" }
}

/**
 * An attribute with default kind, for example `[...]` on line 1 in
 * ```csharp
 * [MyAttribute(0)]
 * int SomeMethod() { return 1; }
 * ```
 */
class DefaultAttribute extends Attribute, @attribute_default {
  override string getAPrimaryQlClass() { result = "DefaultAttribute" }
}

/**
 * A generic attribute with default kind, for example `[...]` on line 1 in
 * ```csharp
 * [MyAttribute<string>(0)]
 * int SomeMethod() { return 1; }
 * ```
 */
class GenericDefaultAttribute extends GenericAttribute, DefaultAttribute {
  override string getAPrimaryQlClass() { result = "GenericDefaultAttribute" }
}

/**
 * An attribute with return kind, for example `[...]` on line 1 in
 * ```csharp
 * [return: MyAttribute(0)]
 * int SomeMethod() { return 1; }
 * ```
 */
class ReturnAttribute extends Attribute, @attribute_return {
  override string toString() { result = "[return: " + getAttributeName(this) + "(...)]" }

  override string getAPrimaryQlClass() { result = "ReturnAttribute" }
}

/**
 * A generic attribute with return kind, for example `[...]` on line 1 in
 * ```csharp
 * [return: MyAttribute<object>(0)]
 * int SomeMethod() { return 1; }
 * ```
 */
class GenericReturnAttribute extends GenericAttribute, ReturnAttribute {
  override string getAPrimaryQlClass() { result = "GenericReturnAttribute" }
}

/**
 * An attribute with assembly kind, for example `[...]` on line 1 in
 * ```csharp
 * [assembly: MyAttribute(0)]
 * ```
 */
class AssemblyAttribute extends Attribute, @attribute_assembly {
  override string toString() { result = "[assembly: " + getAttributeName(this) + "(...)]" }

  override string getAPrimaryQlClass() { result = "AssemblyAttribute" }
}

/**
 * A generic attribute with assembly kind, for example `[...]` on line 1 in
 * ```csharp
 * [assembly: MyAttribute<string>(0)]
 * ```
 */
class GenericAssemblyAttribute extends GenericAttribute, AssemblyAttribute {
  override string getAPrimaryQlClass() { result = "GenericAssemblyAttribute" }
}

/**
 * An attribute with module kind, for example `[...]` on line 1 in
 * ```csharp
 * [module: MyAttribute(0)]
 * ```
 */
class ModuleAttribute extends Attribute, @attribute_module {
  override string toString() { result = "[module: " + getAttributeName(this) + "(...)]" }

  override string getAPrimaryQlClass() { result = "ModuleAttribute" }
}

/**
 * A generic attribute with module kind, for example `[...]` on line 1 in
 * ```csharp
 * [module: MyAttribute<string>(0)]
 * ```
 */
class GenericModuleAttribute extends GenericAttribute, ModuleAttribute {
  override string getAPrimaryQlClass() { result = "GenericModuleAttribute" }
}
