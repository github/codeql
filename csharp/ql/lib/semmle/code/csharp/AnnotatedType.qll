/**
 * Provides predicates and classes for representing annotations on types, such as nullability information.
 *
 * Use an element's `getType()` predicate to access the standard type of the element, without regard
 * for annotations on the type. Use the corresponding `getAnnotatedType()` predicate to
 * get the annotated type, that includes this additional information.
 */

import csharp
private import TypeRef

private module Annotations {
  newtype TAnnotation =
    TReadonlyRefType() or
    TRefType() or
    TOutType() or
    TNullability(@nullability n)

  /** An annotation on a type. */
  class TypeAnnotation extends TAnnotation {
    /** Gets the bit position in the bit-field, also used to order the annotations in the text format. */
    int getBit() { none() }

    /** Gets the string prefixing the type, if any. */
    string getPrefix() { none() }

    /** Gets the string suffixing the type, if any. */
    string getSuffix() { none() }

    /** Gets a string representation of this type annotation. */
    string toString() { none() }
  }

  /** An annotation indicating that the type is a readonly reference. */
  class ReadonlyRefType extends TypeAnnotation, TReadonlyRefType {
    override string getPrefix() { result = "readonly " }

    override string toString() { result = "readonly ref" }

    override int getBit() { result = 4 }
  }

  /** An annotation indicating that the variable or return is by `ref`. */
  class RefTypeAnnotation extends TypeAnnotation, TRefType {
    override string getPrefix() { result = "ref " }

    override string toString() { result = "ref" }

    override int getBit() { result = 5 }
  }

  /** An annotation indicating that the parameter is `out`. */
  class OutType extends TypeAnnotation, TOutType {
    override string getPrefix() { result = "out " }

    override string toString() { result = "out" }

    override int getBit() { result = 6 }
  }

  /**
   * A structured type annotation representing type nullability.
   * For example, `IDictionary<string!,string?>?` has nullability `<!,?>?`.
   */
  class Nullability extends TypeAnnotation, TNullability {
    @nullability nullability;

    Nullability() { this = TNullability(nullability) }

    override string toString() { result = this.getMemberString() + this.getSelfNullability() }

    language[monotonicAggregates]
    private string getMemberString() {
      if nullability_parent(_, _, nullability)
      then
        result =
          "<" +
            concat(int i, Nullability child |
              nullability_parent(getNullability(child), i, nullability)
            |
              child.toString(), "," order by i
            ) + ">"
      else result = ""
    }

    /**
     * Gets a string representing the nullability, disregarding child nullability.
     * For example, `IDictionary<string!,string?>?` has nullability `?`.
     */
    string getSelfNullability() { none() }

    override int getBit() { none() }
  }

  @nullability getNullability(Nullability n) { n = TNullability(result) }

  private newtype TAnnotations =
    TAnnotationFlags(int flags, Nullability n) {
      exists(Element e |
        flags = getElementTypeFlags(e) and
        n = getElementNullability(e)
      )
    } or
    TAnnotationsNoFlags(Nullability n)

  /** A collection of type annotations. */
  class TypeAnnotations extends TAnnotations {
    /** Gets the nullability of this type annotation. */
    Nullability getNullability() {
      this = TAnnotationFlags(_, result)
      or
      this = TAnnotationsNoFlags(result)
    }

    /** Gets text to be displayed before the type. */
    string getTypePrefix() {
      result =
        concat(TypeAnnotation a | a = this.getAnAnnotation() | a.getPrefix(), "" order by a.getBit())
    }

    /** Gets text to be displayed after the type. */
    string getTypeSuffix() {
      result =
        concat(TypeAnnotation a | a = this.getAnAnnotation() | a.getSuffix(), "" order by a.getBit())
    }

    /** Gets a textual representation of this type annotation. */
    string toString() {
      result = this.getTypePrefix() + this.getNullability() + this.getTypeSuffix()
    }

    private int getFlags() { this = TAnnotationFlags(result, _) }

    private predicate isSet(int bit) {
      isBit(bit) and
      exists(int mask | mask = getBitMask(bit) | this.getFlags().bitAnd(mask) = mask)
    }

    /** Gets an annotation in this set of annotations. */
    TypeAnnotation getAnAnnotation() {
      this.isSet(result.getBit())
      or
      result = this.getNullability()
    }
  }

  /** Gets the nullability with no additional flags. */
  Nullability getNoFlagsNullability(TypeAnnotations annotations) {
    annotations = TAnnotationsNoFlags(result)
  }

  /**
   * Gets the `i`th child of type annotations `annotations`.
   * This is used to represent structured datatypes, where the structure
   * of the type annotation mirrors the structure of the annotated type.
   */
  bindingset[i]
  TypeAnnotations getChild(TypeAnnotations annotations, int i) {
    getNoFlagsNullability(result) = getChildNullability(annotations.getNullability(), i)
  }

  /**
   * Gets the `i`th child of nullability `n`.
   * Returns `n` if the nullability is not explicitly
   * stored in the database, since many type annotations will have consistent
   * nullability.
   */
  bindingset[i]
  Nullability getChildNullability(Nullability n, int i) {
    if nullability_parent(_, i, getNullability(n))
    then nullability_parent(getNullability(result), i, getNullability(n))
    else result = n
  }

  /**
   * A type that is "oblivious", either because nullability is not
   * applicable, because the code was not compiled in a nullable context, or
   * because the C# language version is less than 8.
   */
  class ObliviousNullability extends Nullability {
    ObliviousNullability() { nullability instanceof @oblivious }

    override string getSelfNullability() { result = "_" }
  }

  /**
   * A type that is "fully" oblivious. The type itself is oblivious
   * and all type arguments are oblivious.
   */
  class NoNullability extends ObliviousNullability {
    NoNullability() { not nullability_parent(_, _, nullability) }
  }

  /** A type with annotated nullability, `?`. */
  class AnnotatedNullability extends Nullability {
    AnnotatedNullability() { nullability instanceof @annotated }

    override string getSelfNullability() { result = "?" }

    override int getBit() { result = 3 }

    override string getSuffix() { result = "?" }
  }

  /**
   * A ref type not annotated with `?` in a nullable context.
   */
  class NotAnnotatedNullability extends Nullability {
    NotAnnotatedNullability() { nullability instanceof @not_annotated }

    override string getSelfNullability() { result = "!" }

    override int getBit() { result = 2 }

    override string getSuffix() { result = "!" }
  }

  /** Holds if the type annotations `annotations` apply to type `type` on element `element`. */
  predicate elementTypeAnnotations(
    @has_type_annotation element, Type type, TypeAnnotations annotations
  ) {
    (
      annotations = TAnnotationFlags(getElementTypeFlags(element), getElementNullability(element))
      or
      not exists(getElementTypeFlags(element)) and
      getNoFlagsNullability(annotations) = getElementNullability(element)
    ) and
    (
      type = element.(Assignable).getType()
      or
      type = element.(Callable).getReturnType()
      or
      type = element.(Expr).getType()
      or
      type = element.(DelegateType).getReturnType()
      or
      type = element.(FunctionPointerType).getReturnType()
    )
  }
}

/** Holds if `bit` is a valid bit to set in the bitmask. */
private predicate isBit(int bit) { bit = any(Annotations::TypeAnnotation a).getBit() }

private int getBitMask(int bit) {
  isBit(bit) and
  result = 1.bitShiftLeft(bit)
}

private int getElementTypeFlags(@has_type_annotation element) {
  result = strictsum(int b | type_annotation(element, b) | b)
}

private Annotations::Nullability getTypeParameterNullability(
  TypeParameterConstraints constraints, Type type
) {
  if specific_type_parameter_nullability(constraints, getTypeRef(type), _)
  then
    specific_type_parameter_nullability(constraints, getTypeRef(type),
      Annotations::getNullability(result))
  else (
    specific_type_parameter_constraints(constraints, getTypeRef(type)) and
    result instanceof Annotations::NoNullability
  )
}

private Annotations::Nullability getElementNullability(@has_type_annotation element) {
  if type_nullability(element, _)
  then type_nullability(element, Annotations::getNullability(result))
  else result instanceof Annotations::NoNullability
}

private newtype TAnnotatedType =
  TAnnotatedTypeNullability(Type type, Annotations::TypeAnnotations annotations) {
    Annotations::elementTypeAnnotations(_, type, annotations)
    or
    exists(AnnotatedConstructedType c, int i |
      type = c.getType().(ConstructedType).getTypeArgument(i) and
      annotations = Annotations::getChild(c.getAnnotations(), i)
    )
    or
    Annotations::getNoFlagsNullability(annotations) = getTypeParameterNullability(_, type)
    or
    // All types have at least one annotated type
    Annotations::getNoFlagsNullability(annotations) instanceof Annotations::NoNullability
    or
    exists(AnnotatedArrayType at |
      type = at.getType().(ArrayType).getElementType() and
      annotations = Annotations::getChild(at.getAnnotations(), 0)
    )
  }

/** A type with additional type information. */
class AnnotatedType extends TAnnotatedType {
  Type type;
  Annotations::TypeAnnotations annotations;

  AnnotatedType() { this = TAnnotatedTypeNullability(type, annotations) }

  /** Gets a textual representation of this annotated type. */
  string toString() {
    result =
      annotations.getTypePrefix() + this.getUnderlyingType().toStringWithTypes() +
        annotations.getTypeSuffix()
  }

  /** Gets the location of this annotated type. */
  Location getLocation() { result = type.getLocation() }

  /**
   * Gets the unannotated type, for example `string` in `string?`.
   * Note that this might be a nullable value type (`System.Nullable`).
   */
  Type getType() { result = type }

  /**
   * Gets the underlying type, for example `string` in `string?`
   * or `int` in `int?`. This also gets the underlying type of
   * nullable value types (`System.Nullable`).
   */
  final Type getUnderlyingType() {
    if type instanceof NullableType
    then result = type.(NullableType).getUnderlyingType()
    else result = type
  }

  /** Gets the type annotation set of this annotated type. */
  Annotations::TypeAnnotations getAnnotations() { result = annotations }

  /** Gets a type annotation of this annotated type. */
  private Annotations::TypeAnnotation getAnAnnotation() {
    result = this.getAnnotations().getAnAnnotation()
  }

  /** Holds if the type is a non-nullable reference, for example, `string` in a nullable-enabled context. */
  predicate isNonNullableRefType() { this.getAnAnnotation() instanceof Annotations::NoNullability }

  /** Holds if the type is a nullable reference, for example `string?`. */
  predicate isNullableRefType() {
    this.getAnAnnotation() instanceof Annotations::AnnotatedNullability
  }

  /** Holds if the type is a `ref`, for example the return type of `ref int F()`. */
  predicate isRef() { this.getAnAnnotation() instanceof Annotations::RefTypeAnnotation }

  /** Holds if the type is a `ref readonly`, for example the return type of `ref readonly int F()`. */
  predicate isReadonlyRef() { this.getAnAnnotation() instanceof Annotations::ReadonlyRefType }

  /** Holds if the type is an `out`, for example parameter `p` in `void F(out int p)`. */
  predicate isOut() { this.getAnAnnotation() instanceof Annotations::OutType }

  /** Holds if this annotated type applies to element `e`. */
  predicate appliesTo(Element e) { Annotations::elementTypeAnnotations(e, type, annotations) }

  /** Holds if this annotated type is the `i`th type argument of constructed generic 'g'. */
  predicate appliesToTypeArgument(ConstructedGeneric g, int i) {
    Annotations::getNoFlagsNullability(this.getAnnotations()) =
      Annotations::getChildNullability(getElementNullability(g), i) and
    this.getType() = g.getTypeArgument(i)
  }

  /** Holds if this annotated type applies to type parameter constraints `constraints`. */
  predicate appliesToTypeConstraint(TypeParameterConstraints constraints) {
    Annotations::getNoFlagsNullability(this.getAnnotations()) =
      getTypeParameterNullability(constraints, type)
  }
}

/** An array type with additional type information. */
class AnnotatedArrayType extends AnnotatedType {
  override ArrayType type;

  /** Gets the annotated element type of this array, for example `int?` in `int?[]`. */
  final AnnotatedType getElementType() {
    result.getType() = type.getElementType() and
    result.getAnnotations() = Annotations::getChild(this.getAnnotations(), 0)
  }

  private string getDimensionString(AnnotatedType elementType) {
    exists(AnnotatedType et, string res |
      et = this.getElementType() and
      res = type.getArraySuffix() and
      if et.getUnderlyingType() instanceof ArrayType and not et.isNullableRefType()
      then result = res + et.(AnnotatedArrayType).getDimensionString(elementType)
      else (
        result = res and elementType = et
      )
    )
  }

  override string toString() {
    exists(AnnotatedType elementType |
      result =
        annotations.getTypePrefix() + elementType.toString() + this.getDimensionString(elementType) +
          annotations.getTypeSuffix()
    )
  }
}

/** A constructed type with additional type information. */
class AnnotatedConstructedType extends AnnotatedType {
  override ConstructedType type;

  /** Gets the `i`th type argument of this constructed type. */
  AnnotatedType getTypeArgument(int i) {
    result.getType() = type.getTypeArgument(i) and
    result.getAnnotations() = Annotations::getChild(this.getAnnotations(), i)
  }

  override string toString() {
    result =
      annotations.getTypePrefix() + type.getUnboundGeneric().getUndecoratedName() + "<" +
        this.getTypeArgumentsString() + ">" + annotations.getTypeSuffix()
  }

  language[monotonicAggregates]
  private string getTypeArgumentsString() {
    result =
      concat(int i |
        exists(this.getTypeArgument(i))
      |
        this.getTypeArgument(i).toString(), ", " order by i
      )
  }
}
