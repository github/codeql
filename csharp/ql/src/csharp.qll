/**
 * The default C# QL library.
 */

import Customizations
import semmle.code.csharp.Attribute
import semmle.code.csharp.Callable
import semmle.code.csharp.Comments
import semmle.code.csharp.Element
import semmle.code.csharp.Event
import semmle.code.csharp.File
import semmle.code.csharp.Generics
import semmle.code.csharp.Location
import semmle.code.csharp.Member
import semmle.code.csharp.Namespace
import semmle.code.csharp.AnnotatedType
import semmle.code.csharp.Property
import semmle.code.csharp.Stmt
import semmle.code.csharp.Type
import semmle.code.csharp.Using
import semmle.code.csharp.Variable
import semmle.code.csharp.XML
import semmle.code.csharp.Preprocessor
import semmle.code.csharp.exprs.Access
import semmle.code.csharp.exprs.ArithmeticOperation
import semmle.code.csharp.exprs.Assignment
import semmle.code.csharp.exprs.BitwiseOperation
import semmle.code.csharp.exprs.Call
import semmle.code.csharp.exprs.ComparisonOperation
import semmle.code.csharp.exprs.Creation
import semmle.code.csharp.exprs.Dynamic
import semmle.code.csharp.exprs.Expr
import semmle.code.csharp.exprs.Literal
import semmle.code.csharp.exprs.LogicalOperation
import semmle.code.csharp.controlflow.ControlFlowGraph
import semmle.code.csharp.dataflow.DataFlow
import semmle.code.csharp.dataflow.TaintTracking
import semmle.code.csharp.dataflow.SSA

/** Whether the source was extracted without a build command. */
predicate extractionIsStandalone() { exists(SourceFile f | f.extractedStandalone()) }
