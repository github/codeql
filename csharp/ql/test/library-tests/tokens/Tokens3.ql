import csharp

from Variable v
where v.fromSource()
select v
