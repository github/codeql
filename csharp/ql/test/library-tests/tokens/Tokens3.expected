| tokens.cs:7:27:7:33 | param1 |
| tokens.cs:16:13:16:19 | field1 |
| tokens.cs:17:13:17:19 | field2 |
| tokens.cs:18:13:18:18 | field3 |
| tokens.cs:20:34:20:40 | param1 |
| tokens.cs:22:20:22:26 | local1 |
| tokens.cs:23:28:23:34 | local2 |
| tokens.cs:23:38:23:39 | x |
| tokens.cs:24:20:24:25 | local3 |
